### ISSUE TITLE
Timeout issue?

### ISSUE THREAD

**Issue by iohenkies** [matteocorti/check_ssl_cert#98]
Hi,

I've been using the check_ssl_cert plugin in Icinga 2 for some time now without any problems on multiple hosts. Now, as it seems all of a sudden, on 2 of the many hosts I get an error while checking:

`SSL_CERT CRITICAL mydomain.com: No certificate returned`

Checking from the command line:

```
root@check01:~# '/usr/lib/nagios/plugins/check_ssl_cert' '--sni' 'mydomain.com' '-H' 'mydomain.com' '-c' '7' '-p' '443' '-w' '14' -d
[DBG] ROOT_CA = 
[DBG] cURL binary needed. SSL Labs = , OCSP = 1
[DBG] cURL binary not specified
[DBG] cURL available: /usr/bin/curl
expect not available
timeout available (/usr/bin/timeout)
[DBG] perl available: /usr/bin/perl
[DBG] date available: /bin/date
found GNU date with timestamp support: enabling date computations
[DBG] check_ssl_cert version: 1.73.0
[DBG] OpenSSL binary: /usr/bin/openssl
[DBG] OpenSSL version: OpenSSL 1.1.0f  25 May 2017
[DBG] OpenSSL configuration directory: /usr/lib/ssl
[DBG] 0 root certificates installed by default
[DBG] System info: Linux check01 4.9.0-8-amd64 #1 SMP Debian 4.9.110-3+deb9u6 (2018-10-08) x86_64 GNU/Linux
[DBG] Date computation: GNU
[DBG] '/usr/bin/openssl s_client' supports '-servername': using -servername mydomain.com
[DBG] '/usr/bin/openssl s_client' supports '-xmpphost': using -xmpphost mydomain.com
[DBG] temporary file /tmp/check_ssl_certVRNib6 created
[DBG] temporary file /tmp/check_ssl_cert0DRVYK created
[DBG] temporary file /tmp/check_ssl_certaBQxMe created
[DBG] temporary file /tmp/check_ssl_certhrYvdx created
downloading certificate to /tmp
[DBG] mydomain.com is not an IP address
[DBG] executing with timeout (60s): echo 'Q' | /usr/bin/openssl s_client   -connect mydomain.com:443 -servername mydomain.com -showcerts -verify 6     2> /tmp/check_ssl_cert0DRVYK 1> /tmp/check_ssl_certVRNib6
[DBG]   /usr/bin/timeout 60 /bin/sh -c "echo 'Q' | /usr/bin/openssl s_client   -connect mydomain.com:443 -servername mydomain.com -showcerts -verify 6     2> /tmp/check_ssl_cert0DRVYK 1> /tmp/check_ssl_certVRNib6"
[DBG] storing a copy of the retrieved certificate in mydomain.com.crt
[DBG] storing a copy of the OpenSSL errors in mydomain.com.error
Error: verify depth is 6
[DBG] cleaning up temporary files
[DBG]   /tmp/check_ssl_certVRNib6
[DBG]   /tmp/check_ssl_cert0DRVYK
[DBG]   /tmp/check_ssl_certaBQxMe
[DBG]   /tmp/check_ssl_certhrYvdx
SSL_CERT CRITICAL mydomain.com: No certificate returned
```

Manually connecting with:

`/usr/bin/openssl s_client   -connect mydomain.com:443 -servername mydomain.com -showcerts -verify 6`

Takes a while but returns the certificate OK.

Long story short, when I up the timeout to 120 seconds (not usable in Icinga though):

`root@check01:~# '/usr/lib/nagios/plugins/check_ssl_cert' '--sni' 'mydomain.com' '-H' 'mydomain.com' '-c' '7' '-p' '443' '-w' '14' -d -t 120`

The check succeeds. Any ideas why this is happening all of a sudden? On one of these servers there is more than 1 domains running and others are checking out just fine with the plugin.

All servers are fully updated Debian 9.5, all patches, Icinga2 version: r2.9.2-1, check_ssl_cert plugin version 1.73.0 (was 1.72.0 before, same problem).

**Comment by matteocorti** [comment]
It's strange as in case of a timeout the error it should exit differently. Could you try also with `-v` and check the content of `mydomain.com.crt` and `mydomain.com.error`?

**Comment by iohenkies** [comment]
Hi and thanks.
-v returns exactly the same output (?).
mydomain.com.crt is emtpy
mydomain.com.error says 'verify depth is 6'

**Comment by matteocorti** [comment]
Trying with the hostname you posted earlier I get not answer, also with "openssl s_client". The correct error should be a timeout but the issue is openssl not getting the certs.

**Comment by matteocorti** [comment]
There was indeed an error in the handling of the timeout. This solves the plugin problem but not yours. From my machine I never get an answer with hostname you published at the beginning. If you do not want to publish it again we can debug the issue per email.

**Comment by matteocorti** [comment]
Debugged offline. The server is not responding with IPv6 on port 443
